src/MPS.jl

```diff
@@ -541,6 +542,127 @@ function mixed_canonize!(tn::AbstractMPO, orthog_center)
+
+    for i in 1:L
+        contract_ind = inds(ψ; at=Site(i))
+        push!(ψ, tensors(mpo; at=Site(i)))
+        contract!(ψ, contract_ind)
+        merge!(Quantum(ψ).sites, Dict(Site(i) => inds(mpo; at=Site(i))))
+    end
+
+    # Group the parallel bond indices
+    for i in 1:(L - 1)
+        groupinds!(ψ, right_inds[i])
+    end
+
+    if !isnothing(threshold) || !isnothing(maxdim)
+        truncate_sweep!(form(ψ), ψ; threshold, maxdim, normalize)
+    else
+        normalize && normalize!(ψ)
+    end
+
+    return ψ
+end
+
+function evolve!(::MixedCanonical, ψ::AbstractAnsatz, mpo::AbstractMPO; normalize, kwargs...)
+    initial_form = form(ψ)
+    mixed_canonize!(ψ, Site(nsites(ψ))) # We convert all the tensors to left-canonical form
+
+    evolve!(NonCanonical(), ψ, mpo; normalize, kwargs...)
+
+    mixed_canonize!(ψ, initial_form.orthog_center)
+
+    return ψ
+end
+
+function evolve!(::Canonical, ψ::AbstractAnsatz, mpo::AbstractMPO; threshold, maxdim, normalize, kwargs...)
+    # We first join the λs to the Γs to get MixedCanonical(Site(1)) form
+    for i in 1:(nsites(ψ) - 1)
+        contract!(ψ; between=(Site(i), Site(i + 1)), direction=:right)
+    end
+
+    evolve!(NonCanonical(), ψ, mpo; threshold=nothing, maxdim=nothing, normalize=false, kwargs...) # set maxdim and threshold to nothing so we truncate from Canonical form
+
+    if !isnothing(threshold) || !isnothing(maxdim)
+        truncate_sweep!(Canonical(), ψ; threshold, maxdim, normalize)
+    else
+        normalize && canonize!(ψ; normalize)
+    end
+
+    return ψ
+end
+
+function truncate_sweep!(::NonCanonical, ψ::AbstractMPO; threshold, maxdim, normalize)
+    for i in nsites(ψ):-1:2
+        canonize_site!(ψ, Site(i); direction=:left, method=:svd)
+    end
+
+    # left-to-right SVD sweep, get left-canonical tensors and singular values and truncate
+    for i in 1:(nsites(ψ) - 1)
+        canonize_site!(ψ, Site(i); direction=:right, method=:svd)
+
+        (!isnothing(threshold) || !isnothing(maxdim)) &&
+            truncate!(ψ, [Site(i), Site(i + 1)]; threshold, maxdim, normalize, compute_local_svd=false)
+
+        contract!(ψ; between=(Site(i), Site(i + 1)), direction=:right)
+    end
+
+    ψ.form = MixedCanonical(Site(nsites(ψ)))
+
+    return ψ
+end
+
+function truncate_sweep!(::MixedCanonical, ψ::AbstractMPO; threshold, maxdim, normalize)
+    truncate_sweep!(NonCanonical(), ψ; threshold, maxdim, normalize)
+end
+
+function truncate_sweep!(::Canonical, ψ::AbstractMPO; threshold, maxdim, normalize)
+    for i in nsites(ψ):-1:2
+        canonize_site!(ψ, Site(i); direction=:left, method=:svd)
+    end
+
+    # left-to-right SVD sweep, get left-canonical tensors and singular values and truncate
+    for i in 1:(nsites(ψ) - 1)
+        canonize_site!(ψ, Site(i); direction=:right, method=:svd)
+        (!isnothing(threshold) || !isnothing(maxdim)) &&
+            truncate!(ψ, [Site(i), Site(i + 1)]; threshold, maxdim, normalize, compute_local_svd=false)
+    end
```

Comment:
No you can't do this. I specifically created this function because in the `Canonical` form you do not contract the lambdas but still truncate them. This is different from the other methods.

Reply:
okay, i want to better understand this but let's merge it and refactor it later (if we need)